New JS Client Announcement Post

polluterofminds opened the pull request:
@niemyjski please review and make sure I don't sound like an idiot in anything I say here. I want to capture how much work you put into this and why, but I want to make sure it's accurate.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/exceptionless/Website?pullRequest=33) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/exceptionless/Website?pullRequest=33) before we can accept your contribution.<br/><hr/>**Justin Hunter** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/exceptionless/Website?pullRequest=33) it.</sub>

niemyjski commented on content/news/2021/2021-09-01-announcing-the-new-exceptionless-javascript-client.md:
```diff
@@ -0,0 +1,44 @@
+draft: false
+---
+    
+We have had a JavaScript client for several years, but it was quickly becoming out of date. To help address the issues with a fast-moving JavaScript landscape and to extend our JavaScript support, we have recently a completely rewritten client that supports multiple JavaScript frameworks. 
```
Our first client release was in 2015 (https://github.com/exceptionless/Exceptionless.JavaScript/releases/tag/v0.1.0). I wouldn't necessarily say it was out of date, those clients will continue to work and be useful probably for another decade or longer. It just was built before ES6 and before the major browsers supported everything. We built it and could only depend on callbacks and ES3 feature set.  We may want to say this. Our main goals were to modernize the api into what would be idiomatic JavaScript code today. Making use of async and await, fetch and ESM modules (tree shaking). Making the api surface feel like code you would right today. By splitting it out into packages we could build smaller libraries and only consume what you need while allowing us to target specific frameworks.

ejsmith replied:
@niemyjski we are talking about new stuff. We need to sensationalize that as much as possible. I do think we should mention the facts that we are built on TypeScript with strong typing and using async await and fetch. I think it would also be nice to list out the architectural structure where we have a core client that just uses fetch and localstorage for all features. Then we have layers on top of that the provide platform specific things like wiring into different error handlers and having to provide polyfills for things like fetch in nodejs.

ejsmith replied:
I also think it's important that we have at least one image in these articles not only to break things up, but for showing in social media where having an image saying what the tweet is about instead of just text makes a big different for visual scanning through a list of tweets.

ejsmith replied:
![image](https://user-images.githubusercontent.com/282584/132272465-5db925bd-85b5-4db8-a315-d067be4ecb06.png)

I like how @actipro does it. https://twitter.com/Actipro/status/1313115200355262464?s=20

polluterofminds replied:
Agreed. I think playing up the core architecture makes a lot of sense. I'll get an image put together today.

polluterofminds replied:
Lol @ejsmith are you saying my beautiful non-designer design needs work? We'll keep iterating on it :)

ejsmith replied:
Haha, yeah, I know. My attempt would probably be even worse. :-)

polluterofminds replied:
Ok just updated this. Let me know what you think.

niemyjski commented on content/news/2021/2021-09-01-announcing-the-new-exceptionless-javascript-client.md:
```diff
@@ -0,0 +1,44 @@
+
+## Bring the client up to modern JavaScript standards
+
+ECMAScript has progressed a lot in the last ten years. It was clear that we would need to re-write our client to keep up. With our new implementation, it will be much easier to stay on top of new features and experimental features. 
```
We were already using typescript which kept the client somewhat modern. It was just our api surface couldn't be broken without a major bump, and without that we couldn't use modern features. 
We may want to talk about some of the new features here. 

- Like ESM solves the issues we had issues with module bundling. We had to have a special umd bundle which supported universal apps (browser and node) and this caused a bunch of issues.
-  We also had a lot of legacy JS usages (callbacks, and IE6 support) which prevented us from simplifying api usages for plugins. Was also harder to support Promises for things like getting source maps and modern submission methods like fetch. 
- We are also now using modern browser api's to pause the client when not in focus/active window so the page can go to sleep easier and thus longer battery life

niemyjski replied:
Also not sure if it's worth comparing code wise some of the old to the new usages, to show off how much easier it is.

niemyjski commented on content/news/2021/2021-09-01-announcing-the-new-exceptionless-javascript-client.md:
```diff
@@ -0,0 +1,44 @@
+
+## Have a little more code separation
+
+This is not a question of monorepo or not. Instead, we recognized that we were trying to cover browser code and server code all in one package. This made things difficult, not to mention we didn't yet have support for specific frameworks (besides Angular 1.0). 
```
Should we link to monorepo npm docs?

ejsmith replied:
No, I don't think then we need to point out that kind of thing in an announcement post. This is about us. I'd rather see some links to our code.

niemyjski commented on content/news/2021/2021-09-01-announcing-the-new-exceptionless-javascript-client.md:
```diff
@@ -0,0 +1,44 @@
+
+This is not a question of monorepo or not. Instead, we recognized that we were trying to cover browser code and server code all in one package. This made things difficult, not to mention we didn't yet have support for specific frameworks (besides Angular 1.0). 
+
+Since we knew we were going to extend our support, it made sense to rebuild the client in a way that allowed us to publish updates to specific sub-modules rather than to the entire core JavaScript client every time. 
```
We currently are not doing this but definitely can and will in the future.

polluterofminds replied:
I was trying to point out that the core package drives all the other ones. I'll try to make that more clear, especially when talking about the architecture earlier in the post.

niemyjski commented:
Added some comments, I like the positivity of this post :)

polluterofminds commented:
@niemyjski @ejsmith I just pushed up some updates based on the comments. Let me know what you think. I'm trying to balance the announcement vibe of this post with the desire to share a lot of technical details. I don't think this post should go very deep technically, but I did add more, including a couple images.

niemyjski commented:
Looks good, we should just change the `content/news/2021/Announcement.png`  image path to be a bit more specific.

polluterofminds commented:
@niemyjski @ejsmith I updated the path for the announcement image. I also updated the date for the post. When this is ready to merge, let me know because I'll also put it out on Twitter for you guys.

ejsmith commented:
Just pushed a commit working to remove the SEO plugin. Some of the meta tags aren't correct yet. Hard to test because the site won't build on Windows. Title doesn't have `- Exceptionless` and image needs to be the fully qualified URL.
